### PR TITLE
[Feature] Implementation of Prompt Logprobs support for TPU

### PR DESCRIPTION
**Pull request by sierraisland**
# Description

This PR introduces comprehensive support for Prompt Logprobs on TPU, allowing users to retrieve log probabilities for both prompt tokens and generated tokens. The implementation handles the complex memory layout requirements of TPU SPMD alignment, Data Parallelism, and Asynchronous Scheduling.

Related bug: b/508315177

# Tests

Test results matched for both DP and non DP.

Results replay: 
- `python tpu-inference/examples/rl_inference.py --model meta-llama/Llama-3.2-1B-Instruct --data-parallel-size 1 --prompt-logprobs 3`
- `python tpu-inference/examples/rl_inference.py --model meta-llama/Llama-3.2-1B-Instruct --data-parallel-size 4 --prompt-logprobs 3`
- `python tpu-inference/examples/offline_inference.py --model meta-llama/Llama-3.2-1B-Instruct --data-parallel-size 1`
- `python tpu-inference/examples/offline_inference.py --model meta-llama/Llama-3.2-1B-Instruct --data-parallel-size 4`

Current version also works for `python tpu-inference/examples/rl_inference.py --data-parallel-size 4 --mixed-logprobs`

Passed tests:
- `pytest tpu-inference/tests/runner/test_tpu_runner.py`
- `pytest tpu-inference/tests/runner/test_tpu_runner_dp.py`
- `pytest tpu-inference/tests/layers/jax/sample/test_sampling.py`
- `pytest tpu-inference/tests/core/test_dp_scheduler.py`
- `pytest tpu-inference/tests/runner/test_structured_decoding_manager.py`
- `pytest tpu-inference/tests/runner/test_speculative_decoding_manager.py`
- `pytest tpu-inference/tests/spec_decode/test_eagle3.py`
- `pytest tpu-inference/tests/runner/test_utils.py`
- `pytest tpu-inference/tests/runner/test_input_batch.py`
- `pytest tpu-inference/tests/runner/test_kv_cache_manager.py`
- `pytest -s -v tpu-inference/tests/e2e/test_speculative_decoding.py`

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/runner/tpu_runner.py [hunk 1]
```diff
@@ -241,14 +261,235 @@ def _jax_logprobs_materialize(
         logprobs_arr = logprobs_arr[logits_indices_selector]
         selected_token_ranks = selected_token_ranks[logits_indices_selector]
 
-    return LogprobsLists(
-        logprob_token_ids=np.array(log_token_ids.tolist()),
-        logprobs=np.array(logprobs_arr.tolist()),
-        sampled_token_ranks=np.array(selected_token_ranks.tolist()),
+    # Use CPU tensors for V1 serialization
+    return LogprobsTensors(
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py:300]: why do we need to change the output type from numpy to torch?

File: tpu_inference/runner/tpu_runner.py [hunk 2]
```diff
@@ -1110,15 +1404,18 @@ def _sample_from_logits(
                 logprobs_tensors=logprobs,
                 expert_indices=expert_indices,
                 total_num_scheduled_tokens=scheduler_output.
-                total_num_scheduled_tokens)
+                total_num_scheduled_tokens,
+                num_logits_per_req=num_logits_per_req,
+                any_prompt_logprobs=any_prompt_logprobs,
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: would it be possible to not mix custom fields (e.g. any_prompt_logprobs) with expected fields in `ModelRunnerOutput?` if we must add these custom fields, can we add an underscore to distinguish them from the expected fields? e.g. `_any_prompt_logprobs`

File: tpu_inference/runner/tpu_runner.py [hunk 3]
```diff
@@ -1110,15 +1404,18 @@ def _sample_from_logits(
...
+                sampling_indices_selector=sampling_indices_selector,
+                gi_physical_indices=gi_physical_indices)
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: curious what does `gi` stand for?

File: tpu_inference/runner/tpu_runner.py [hunk 4]
```diff
@@ -122,22 +127,30 @@ def __init__(
...
-            next_tokens_cpu = next_tokens_cpu[self.logits_indices_selector]
-        selected_token_ids = np.expand_dims(next_tokens_cpu[:self._num_reqs],
-                                            1)
+
+        # Use the dedicated sampling selector to unsuffle tokens from multiple chips
+        # Only use the portion corresponding to the current active requests
+        active_selector = self.sampling_indices_selector[:self._num_reqs]
+        selected_token_ids = next_tokens_cpu[active_selector]
+        selected_token_ids = np.expand_dims(selected_token_ids, 1)
+
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: why is this change needed?

File: tpu_inference/runner/tpu_runner.py [hunk 5]
```diff
@@ -122,22 +127,30 @@ def __init__(
...
         self._model_runner_output.sampled_token_ids = valid_sampled_token_ids
 
         if self._logprobs_tensors is not None:
-            # Use materialize to ensure logprobs are ready on host when we return async results
-            self._model_runner_output.logprobs = _jax_logprobs_materialize(
-                self._logprobs_tensors, self.logits_indices_selector)
+            # Separate prompt logprobs and sampling logprobs
+            self._model_runner_output.logprobs, self._model_runner_output.prompt_logprobs_dict = _separate_logprobs(
+                self._logprobs_tensors, self.logits_indices_selector,
+                self.num_logits_per_req, self._model_runner_output.req_ids)
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: would it be possible to do something like the following for clarity?

```
 self._model_runner_output.logprobs = _jax_logprobs_materialize(
                self._logprobs_tensors, self.logits_indices_selector)
```

and then 
```
self._model_runner_output.logprobs, self._model_runner_output.prompt_logprobs_dict = _separate_logprobs(self._model_runner_output.logprobs, self.num_logits_per_req, self._model_runner_output.req_ids)
```

File: tpu_inference/runner/tpu_runner.py [hunk 6]
```diff
@@ -1256,22 +1554,22 @@ def _prepare_dp_input_metadata(self,
             self.num_reqs_paddings_per_dp, max_num_reqs_across_dp)
         padded_num_reqs = padded_num_reqs_per_dp_rank * dp_size
 
-        all_req_indices = np.concatenate(
-            [req_indices_dp[dp_rank] for dp_rank in range(dp_size)])
-        all_positions = np.concatenate([
-            np.arange(len(req_indices_dp[dp_rank])) +
-            padded_num_reqs_per_dp_rank * dp_rank for dp_rank in range(dp_size)
-        ])
+        # sampling_indices_selector: maps [0..num_reqs-1] -> [0..padded_num_reqs-1]
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py:1743]: can you help me understand this logic? why is `logits_indices_selector` being replaced by `sampling_indices_selector`?

File: tpu_inference/runner/tpu_runner.py [hunk 7]
```diff
@@ -1360,12 +1658,32 @@ def _apply_async_token_substitution(self, input_ids,
         return input_ids
 
     def _prepare_inputs(self, scheduler_output: "VllmSchedulerOutput"):
+        use_spec_decode = len(
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py:1923]: does prompt logs work with speculative decoding?

File: tpu_inference/runner/tpu_runner.py [hunk 8]
```diff
@@ -1360,12 +1658,32 @@ def _apply_async_token_substitution(self, input_ids,
...
+            return self._prepare_inputs_non_dp(scheduler_output,
+                                               use_spec_decode,
+                                               any_prompt_logprobs)
+
+    def _prepare_inputs_dp(self,
+                           scheduler_output: "VllmSchedulerOutput",
+                           use_spec_decode: Optional[bool] = None,
+                           any_prompt_logprobs: Optional[bool] = None):
+        if use_spec_decode is None:
+            use_spec_decode = len(
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: this `use_spec_decode` recalculation might be redundant.

File: tpu_inference/runner/tpu_runner.py [hunk 9]
```diff
@@ -1360,12 +1658,32 @@ def _apply_async_token_substitution(self, input_ids,
...
+                scheduler_output.scheduled_spec_decode_tokens) > 0
+        if any_prompt_logprobs is None:
+            any_prompt_logprobs = any(
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: similar for `any_prompt_logprobs` since it was already compued in  `prepare_inputs`

File: tpu_inference/runner/tpu_runner.py [hunk 10]
```diff
@@ -241,14 +261,235 @@ def _jax_logprobs_materialize(
...
+) -> int:
+    """Calculates the total number of logits (sampled tokens) to be returned.
+
+    In DP mode, we must ensure each rank has an identical-sized shard to 
+    maintain SPMD alignment and avoid memory overwrites between ranks.
+    """
+    if not (use_spec_decode or any_prompt_logprobs):
+        return -1
+
+    if dp_size <= 1:
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: can we update this to `==1`?

File: tpu_inference/runner/tpu_runner.py [hunk 11]
```diff
@@ -241,14 +261,235 @@ def _jax_logprobs_materialize(
...
+    num_draft_tokens = np.zeros(num_reqs, dtype=np.int32)
+    for req_id, draft_token_ids in (
+            scheduler_output.scheduled_spec_decode_tokens.items()):
+        if req_id in req_id_to_index:
+            req_idx = req_id_to_index[req_id]
+            num_draft_tokens[req_idx] = len(draft_token_ids)
+    return num_draft_tokens
+
+
+def _get_total_sampled_tokens(
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: I think some of the logic in this function might be replicating what `_prepare_dp_input_metadata` does. would it be possible to simply the code by resolving some of the redundancy?

File: tpu_inference/runner/tpu_runner.py [hunk 12]
```diff
@@ -1026,8 +1299,26 @@ def _sample_from_logits(
...
+                    tokens_1d = jnp.reshape(next_tokens, (-1, ))
+
+                    padded_next_tokens = jnp.zeros((logits.shape[0], ),
+                                                   dtype=next_tokens.dtype)
+                    padded_next_tokens = padded_next_tokens.at[:tokens_1d.
+                                                               shape[0]].set(
+                                                                   tokens_1d)
+                    token_ids_for_gather = padded_next_tokens
+                else:
+                    token_ids_for_gather = next_tokens
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: can you help me understand what this logic is doing? 
is it masking out all the padding tokens in this batch?

File: tpu_inference/runner/tpu_runner.py [hunk 13]
```diff
@@ -241,14 +261,235 @@ def _jax_logprobs_materialize(
...
+    padded_num_reqs_per_dp_rank: int,
+    any_prompt_logprobs: bool,
+    use_spec_decode: bool,
+) -> Tuple[List[int], List[int], np.ndarray]:
+    """Standard Interleaved Layout. Ensures compatibility with Async Token Substitution."""
+    logits_indices_selector = []
+    num_logits_per_req = []
+    is_dp = int(dp_size) > 1
+    padded_num_reqs = int(padded_num_reqs_per_dp_rank) * int(dp_size)
+    gi_physical_indices = np.zeros(padded_num_reqs, dtype=np.int32)
```
Comment (wenxindongwork) [inline tpu_inference/runner/tpu_runner.py]: what is the final value of gi_physical_indices and what does it stand for?

File: tpu_inference/runner/tpu_runner.py (hunk 1 above)
Comment (sierraisland) [reply tpu_inference/runner/tpu_runner.py:300]: LogprobsTensors required to use torch.

File: tpu_inference/runner/tpu_runner.py (hunk 4 above)
Comment (sierraisland) [reply tpu_inference/runner/tpu_runner.py]: Because use sampling_indices_selector instead of logits_indices_selector.
- logits_indices_selector: Maps flattened device logprobs back to individual requests (1-to-many).
- sampling_indices_selector: Maps the exact next-token generation slots (1-to-1).

File: tpu_inference/runner/tpu_runner.py (hunk 7 above)
Comment (sierraisland) [reply tpu_inference/runner/tpu_runner.py:1923]: test tpu-inference/tests/e2e/test_speculative_decoding.py passed

File: tpu_inference/runner/tpu_runner.py (hunk 6 above)
Comment (sierraisland) [reply tpu_inference/runner/tpu_runner.py:1743]: In Data Parallel mode, requests are split across multiple TPU chips. When JAX returns the sampled tokens, they are concatenated in "rank-order" (all tokens from Rank 0, then Rank 1, etc.), which doesn't match the original request order expected by vLLM. It maps the original request index [0..num_reqs-1] to its physical location in the padded JAX output [0..padded_num_reqs-1]. logits_indices_selector and sampling_indices_selector answered above.

**Review by wenxindongwork — COMMENTED** [review]
before fixing the issues in this PR, can you take a look at my implementation in 

https://github.com/vllm-project/tpu-inference/tree/wxd-prompt-logprobs ?

There are a couple issues worth pointing out 

1. currently the logic is getting unnecessarily complicated in this PR. I would recommend landing a first implementation assuming that all requests require prompt logprobs, which helps to simplify the logic. 
2. I believe you will need to keep track of prompt_logprobs in a cached request state for supporting chunk prefill. 
3. I compared the output of my branch vs GPU numbers using offline_inference.py, it looks good to me, but worth double checking with longer prompt length to trigger chunk prefill.
